### ISSUE TITLE
Install in Tomcat in a Virtual Machine

### ISSUE THREAD

**Issue by talesrands** [opendatakit/aggregate#201]
Hello,

I'm using a Mac and trying to use ODK Aggregate in a Virtual Machine (Ubuntu), through Tomcat and PostgreSQL.

When downloading the ODK Aggregate I used the OSX Installer and then sent the ODKAggregate.war generated in my Mac to my VM, an Ubuntu, where the Tomcat and the PostgreSQL are installed.

The commands in create_db_and_user.sql were passed to the VM's terminal line-by-line, and ran without any kind of feedbacks (nor errors).

The Tomcat is not running the ODKAggregate, and the server:8080/ODKAggregate gives me Error 404. Yes, I've waited several minutes and tried again. Is it caused by the difference of OS or should I look for other kinds of possible errors?

**Comment by ggalmazor** [comment]
Hi @talesrands!

What you did should work, I think.

Let's discard the easy ones first: Aggregate must be deployed as the ROOT app in Tomcat. Stop Tomcat, remove any files inside `webapps/ROOT` and unzip there the contents of the WAR file and start Tomcat again.

If it doesn't work, please, copy here the contents of `catalina.out` log file.

**Comment by ggalmazor** [comment]
Hi @talesrands! @wbrunette has made me aware of something I didn't know. It turns out that Aggregate **can be deployed** in any directory. No need to deploy it as the ROOT application on Tomcat8. Sorry for the confusion!

We need to be sure that Tomcat is, in fact, deploying Aggregate. Could you tell how you sent Aggregate to Tomcat? Was it with the manager app? Did you copy the WAR to webapps?

Also, the `catalina.out` log file would help us better understand what's going on...

**Comment by talesrands** [comment]
Hello @ggalmazor . Thanks for helping me.

I copied the WAR to webapps.

So, I look at ``catalina.out`` and the errors were about password authentication.  While revisiting my steps I found out there was a problem when I was preparing the postgreSQL. When I passed the command ``\i create_db_and_user.sql`` the following error occurred:
> psql:create_db_and_user.sql:2: ERROR:  UNENCRYPTED PASSWORD is no longer supported
> LINE 1: create user "myuser" with unencrypted password 'mypassword';

I don't have a SQL Certificate, so that might be it. Is it the problem? The installer had the option "I don't have an SQL Certificate", which is weird if that's not really an option.

I'm using Tomcat8.

**Comment by ggalmazor** [comment]
Oh, now I guess that the problem comes from Aggregate trying to connect to your PostgreSQL. What version do you install? Aggregate requires PostgreSQL 9.6.

We use non-encrypted database credentials and I'm reading that support for that was dropped with PostgreSQL 10.

**Comment by talesrands** [comment]
When I use the command line ``psql`` I see this: ``psql (10.2 (Ubuntu 10.2-1.pgdg16.04+1))``.

I guess that means I have the version 10.2, right?

So, I did the following:

``talesrands@ssdnodes-XXXXX:~/testes/ODKAggregate$ sudo apt-get remove postgresql``
> Reading package lists... Done
> Building dependency tree       
> Reading state information... Done
> The following packages were automatically installed and are no longer required:
> linux-headers-4.4.0-62 linux-headers-4.4.0-62-generic linux-image-4.4.0-62-generic linux-image- extra-4.4.0-62-generic
> Use 'sudo apt autoremove' to remove them.
>The following packages will be REMOVED:
  postgresql
0 upgraded, 0 newly installed, 1 to remove and 33 not upgraded.
After this operation, 63.5 kB disk space will be freed.
Do you want to continue? [Y/n] Y
(Reading database ... 164995 files and directories currently installed.)
Removing postgresql (10+190.pgdg16.04+1) ...

Then I assumed it was uninstalled and tried to specify the version I wanted. My next command was:

> talesrands@ssdnodes-61562:~/testes/ODKAggregate$ sudo apt-get install postgresql-9.6
> Reading package lists... Done
>Building dependency tree       
>Reading state information... Done
> postgresql-9.6 is already the newest version (9.6.7-1.pgdg16.04+1).
> The following packages were automatically installed and are no longer required:
  linux-headers-4.4.0-62 linux-headers-4.4.0-62-generic linux-image-4.4.0-62-generic linux-image->extra-4.4.0-62-generic
>Use 'sudo apt autoremove' to remove them.
>0 upgraded, 0 newly installed, 0 to remove and 33 not upgraded.

But when I did a ``psql`` again, I had the same return as before: ``psql (10.2 (Ubuntu 10.2-1.pgdg16.04+1))``

**Comment by ggalmazor** [comment]
Maybe you still have the client package from version 10. If you do a `sudo dpkg -l | grep postgresql` do you see the `postgresql-client` package for version 9.6?

**Comment by talesrands** [comment]
I see for both versions now. 

>ii  pgdg-keyring                        2017.3                                     all          keyring for apt.postgresql.org
ii  postgresql-10                       10.2-1.pgdg16.04+1                         amd64        object-relational SQL database, version 10 server
ii  postgresql-9.6                      9.6.7-1.pgdg16.04+1                        amd64        object-relational SQL database, version 9.6 server
ii  postgresql-client-10                10.2-1.pgdg16.04+1                         amd64        front-end programs for PostgreSQL 10
ii  postgresql-client-9.6               9.6.7-1.pgdg16.04+1                        amd64        front-end programs for PostgreSQL 9.6
ii  postgresql-client-common            190.pgdg16.04+1                            all          manager for multiple PostgreSQL client versions
ii  postgresql-common                   190.pgdg16.04+1                            all          PostgreSQL database-cluster manager
ii  postgresql-contrib-9.6              9.6.7-1.pgdg16.04+1                        amd64        additional facilities for PostgreSQL

**Comment by ggalmazor** [comment]
Ok, try removing any `postgresql-*-10` you have found. Maybe you need to purge apt and reboot... I'm not sure...

**Comment by talesrands** [comment]
I removed two files I found: ``/etc/postgresql/10`` and ``/var/lib/postgresql/10``. 

The folders ``/etc/postgresql/9.6`` and ``/var/lib/postgresql/9.6`` are still there.

Now it happens as follows:
``talesrands@ssdnodes-XXXXX:/var/lib/postgresql$ psql``
>psql: FATAL:  role "talesrands" does not exist
``talesrands@ssdnodes-61562:/var/lib/postgresql$ psql --username postgres``
>psql: FATAL:  Peer authentication failed for user "postgres"

**Comment by ggalmazor** [comment]
Try with `sudo su postgres` and then execute `psql`

**Comment by talesrands** [comment]
It actually worked, and the message changed. Now it looks like this:
> psql (10.2 (Ubuntu 10.2-1.pgdg16.04+1), server 9.6.7)

Does that mean it worked? Will I always have to do the ``sudo su postgres`` ?

**Comment by ggalmazor** [comment]
No, just to prepare the database for Aggregate. I think you need to follow these steps again:
https://github.com/opendatakit/aggregate#local-postgresql-server

Then we need to restart Tomcat and see if Aggregate starts correctly

**Comment by talesrands** [comment]
I used the commands:
``postgres=# \cd /home/talesrands/testes/ODKAggregate``
``postgres=# \i create_db_and_user.sql``
>CREATE DATABASE
CREATE ROLE
GRANT
ALTER DATABASE
psql (10.2 (Ubuntu 10.2-1.pgdg16.04+1), server 9.6.7)
You are now connected to database "dtm_name" as user "postgres".
CREATE SCHEMA
GRANT
ALTER SCHEMA

Then I moved the corresponding ``ODKAggregate.war`` to the ``webapps`` folder, aaaaaand it is still not working. However, I opened catalina.out and the errors now have changed. Now it says:

>Connection to 127.0.0.1:5432 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP conections.

That's probably a different problem, and the one of this topic is solved. For that, thank you very much. I'd never think it was about the PostgreSQL version.

For what it's worthy, I think I caused this new problem during the usage of the ODK Installer to build the WAR file. When asked for the "Database server hostname" there was a Hint on the screen saying "If the database server is in the same host as the webserver, enter 127.0.0.1", and I wasn't sure what 'webserver' it was talking about so I used "127.0.0.1".

**Comment by ggalmazor** [comment]
Ok, I'm glad we've solved the database issue :)

If postgresql and tomcat are installed on the same host, try running the installer again and this time enter "localhost" as database server host. Then clean webapps and move the new WAR. To be safe, stop the tomcat before doing this.

**Comment by talesrands** [comment]
I used the following command: ``sudo netstat -plunt |grep postgres``, and got the following:
>tcp        0      0 127.0.0.1:5433          0.0.0.0:*               LISTEN      1398/postgres   
tcp6       0      0 ::1:5433                :::*                    LISTEN      1398/postgres   

I think the PostgreSQL 9.6 used port 5433 because the 5432 was occupied by the PostgreSQL 10 and probably that was the problem. I changed the port in ``postgresql.conf`` and it worked! Thank you very much for your help!
